trace/trace.go

```diff
@@ -123,73 +107,83 @@ func WithSpan(parent context.Context, s *Span) context.Context {
 	return context.WithValue(parent, contextKey{}, s)
 }
 
-// StartSpanOptions contains options concerning how a span is started.
-type StartSpanOptions struct {
+// StartOptions contains options concerning how a span is started.
+type StartOptions struct {
 	// RecordEvents indicates whether to record data for this span, and include
 	// the span in a local span store.
 	// Events will also be recorded if the span will be exported.
 	RecordEvents bool
-	Sampler      // if non-nil, the Sampler to consult for this span.
+
+	Sampler // if non-nil, the Sampler to consult for this span.
+
 	// RegisterNameForLocalSpanStore indicates that a local span store for spans
 	// of this name should be created, if one does not exist.
 	// If RecordEvents is false, this option has no effect.
 	RegisterNameForLocalSpanStore bool
+
+	// Tags are the tags applied to the span. Optional.
+	Tags *tag.Map
 }
 
 // StartSpan starts a new child span of the current span in the context.
 //
 // If there is no span in the context, creates a new trace and span.
-func StartSpan(ctx context.Context, name string) context.Context {
+func StartSpan(ctx context.Context, name string) (context.Context, *Span) {
 	parentSpan, _ := ctx.Value(contextKey{}).(*Span)
-	return WithSpan(ctx, parentSpan.StartSpanWithOptions(name, StartSpanOptions{}))
+	span := NewSpan(name, parentSpan, StartOptions{})
+	return WithSpan(ctx, span), span
 }
 
 // StartSpanWithOptions starts a new child span of the current span in the context.
 //
 // If there is no span in the context, creates a new trace and span.
-func StartSpanWithOptions(ctx context.Context, name string, o StartSpanOptions) context.Context {
+func StartSpanWithOptions(ctx context.Context, name string, o StartOptions) (context.Context, *Span) {
 	parentSpan, _ := ctx.Value(contextKey{}).(*Span)
-	return WithSpan(ctx, parentSpan.StartSpanWithOptions(name, o))
+	span := NewSpan(name, parentSpan, o)
+	return WithSpan(ctx, span), span
 }
 
-// StartSpanWithRemoteParent starts a new child span with the given parent SpanContext.
-//
-// If there is an existing span in ctx, it is ignored -- the returned Span is a
-// child of the span specified by parent.
-func StartSpanWithRemoteParent(ctx context.Context, name string, parent SpanContext, o StartSpanOptions) context.Context {
-	return WithSpan(ctx, NewSpanWithRemoteParent(name, parent, o))
-}
-
-// StartSpan starts a new child span.
+// EndSpan ends the current span.
 //
-// If s is nil, creates a new trace and span, like the function NewSpan.
-func (s *Span) StartSpan(name string) *Span {
-	return s.StartSpanWithOptions(name, StartSpanOptions{})
+// The context passed to EndSpan will still refer to the now-ended span, so any
+// code that adds more information to it, like SetSpanStatus, should be called
+// before EndSpan.
+func EndSpan(ctx context.Context) {
+	s, ok := ctx.Value(contextKey{}).(*Span)
+	if !ok {
+		return
+	}
+	s.End(EndOptions{Tags: tag.FromContext(ctx)})
```

Comment:
This extra bit does seem like it would be nice to have EndSpan for convenience.

Reply:
But I see below that you may defer tags until later, in which case I agree this can be removed.